src/python/pants/engine/internals/build_files.py

```diff
@@ -169,6 +173,61 @@ async def ensure_address_family(request: OptionalAddressFamily) -> AddressFamily
     return request.ensure()
 
 
+@dataclass(frozen=True)
+class BUILDFileEnvironmentVariablesRequest:
+    file_content: FileContent
+
+
+class BUILDFileEnvVarExtractor(ast.NodeVisitor):
+    def __init__(self):
+        super().__init__()
+        self.env_vars: set[str] = set()
+
+    @classmethod
+    def get_env_vars(cls, file_content: FileContent) -> Sequence[str]:
+        obj = cls()
+        obj.visit(ast.parse(file_content.content, file_content.path))
+        return tuple(obj.env_vars)
+
+    def visit_Call(self, node: ast.Call):
+        is_env = isinstance(node.func, ast.Name) and node.func.id == "env"
+        for arg in node.args:
+            if not is_env:
+                self.visit(arg)
+                continue
+
+            # Only first arg may be checked as env name
+            is_env = False
+
+            if sys.version_info[0:2] < (3, 8):
+                value = arg.s if isinstance(arg, ast.Str) else None
+            else:
+                value = arg.value if isinstance(arg, ast.Constant) else None
+            if value:
+                # Found env name in this call, we're done here.
+                self.env_vars.add(value)
+                return
```

Comment:
This is pretty conservative, which is nice: but it might be a good idea to log a warning if the node wasn't of a known type (in case someone tries to compute an env var, and then wonders why it fails / doesn't find anything).